table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/CompatibleMetastoreClient.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.table.under.hive.util;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.thrift.TApplicationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+class CompatibleMetastoreClient implements InvocationHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(CompatibleMetastoreClient.class);
+
+  private final IMetaStoreClient mDelegate;
+  private final HMSShim mCompat;
+
+  CompatibleMetastoreClient(IMetaStoreClient delegate, HMSShim compatibility) {
+    mDelegate = delegate;
+    mCompat = compatibility;
+  }
+
+  @Override
+  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+    try {
+      return method.invoke(mDelegate, args);
+    } catch (InvocationTargetException delegateException) {
+      try {
+        LOG.info("Couldn't invoke method {}", method.toGenericString());
+        if (mCompat != null
+            && delegateException.getCause().getClass()
+            .isAssignableFrom(TApplicationException.class)) {
+          LOG.info("Attempting to invoke with {}", mCompat.getClass().getName());
```

Comment:
These info logs could be spammy, since it would log for each invocation of those calls. Should we just cache this info and proactively use the compat method since that decision won't change over time. Then, we can log it when cache the method decision.

Reply:
reduced log level